### DISCUSSION TITLE
PickList security issue : on validators phase of JSF life cycle - no data validation ? Consequence : no message "Validation Error: Value is not valid" when invalid values are submitted

### DISCUSSION THREAD

**Discussion by evguen** [primefaces/community#1490] [PrimeFaces] [answered]
Hello,

Configuration : PrimeFaces 12.0.3, JSF, MyFaces 3.0.0

When a form is submitted, the components passes through validators phase of JSF life cycle. 

**Example #1 - with MyFaces selectOneMenu component**

For instance, when rendered on a webpage a listbox contains 3 values : 

```xhtml
<h:selectOneMenu value="#{page.selectedFruit}" id="fruit">
    <f:selectItem itemLabel="Apple" itemValue="1" />
    <f:selectItem itemLabel="Banana" itemValue="2" />
    <f:selectItem itemLabel="Pear" itemValue="3" />
</h:selectOneMenu>
```

the rendered HTML will be : 

```html
<select id="fruit" name="fruit" size="1">	
  <option value="1">Apple</option>	
  <option value="2">Banana</option>	
  <option value="3">Pear</option>
</select>
```

When the form is submitted, if we tamper the HTTP request to set a value which is outside of the possible values (`1`, `2` or `3`) by submitting another value, for example `10` - the JSF validation component will throw an error `fruit: Validation Error: Value is not valid` - because the value `10` was not initially available. This error is thrown by `jakarta.faces.component.UISelectOne#processValidators`.

**Example #2 - with PrimeFaces selectOneMenu component**

This validation is also available when using PrimeFaces components like 

```xhtml
<p:selectOneMenu value="#{page.selectedFruit}">
    <f:selectItem itemLabel="Apple" itemValue="1" />
    <f:selectItem itemLabel="Banana" itemValue="2" />
    <f:selectItem itemLabel="Pear" itemValue="3" />
</p:selectOneMenu>
```

The validator called is implemented in `org.primefaces.component.selectonemenu.SelectOneMenu#validateValue` wich will call `super.validateValue(context, value);` (which is `jakarta.faces.component.UISelectOne#processValidators` - same as the first example)

**Example #3 - with PrimeFaces pickList component**

Inspired from : http://www.primefaces.org:8080/showcase/ui/data/pickList.xhtml

XHTML : 
```xhtml
    <p:pickList id="pickList" value="#{frag.cities}" var="cities" itemLabel="#{cities}" itemValue="#{cities}"/>
```

Java : 
```java
@Named
@RequestScoped
public class PickListView {
    private DualListModel<String> cities;

    @PostConstruct
    public void init() {
        //Cities
        List<String> citiesSource = new ArrayList<>();
        List<String> citiesTarget = new ArrayList<>();

        citiesSource.add("San Francisco");
        citiesSource.add("London");
        citiesSource.add("Paris");
        citiesSource.add("Istanbul");
        citiesSource.add("Berlin");
        citiesSource.add("Barcelona");
        citiesSource.add("Rome");

        cities = new DualListModel<>(citiesSource, citiesTarget);

    }

    public DualListModel<String> getCities() {
        return cities;
    }

    public void setCities(DualListModel<String> cities) {
        this.cities = cities;
    }
}
```

On the web page I select and move to the right column "San Francisco" and "London". Then I tamper the HTTP request (or modify the HTML source code within the browser before submitting the form) to modify all occurrences of "London" by "Petrus" and submit the form. 

The java `cities` will contain "San Francisco" and "Petrus".

Same behavious with more complex objects with `converters`.

I beleive there should be a validator to verify that submited values are from the initial list. 

From my point of view, this is kind a security issue for users relying on validators phase of JSF life cycle as it works for many other components.

Could you please share your point of view and confirm is this is an issue ? 

Best Regards

**Comment by melloware** [comment]
Excellent diagnosis.  I tend to agree with your assessment if it is not currently working like other JSF components what do you think @tandraschko @Rapster ?

**Reply by evguen** [reply]
Hello,
may I kindly remind you about this issue ?
Have a nice day. 
Best Regards

**Reply by tandraschko** [reply]
i currently dont have time for it in my sparetime, sorry.

**Reply by melloware** [reply]
@evguen you should open a PrimeFaces GitHub issue if you want this tracked with an executable example using the [PrimeFaces Test](https://github.com/primefaces/primefaces-test) project.

**Reply by Del-Tab** [reply, marked as answer]
@melloware as you noticed, I opened https://github.com/primefaces/primefaces/issues/12059 with an example from primefaces-test  project :)

**Reply by melloware** [reply]
14.0.1 is now in Maven Central
